packages/kbn-alerts-ui-shared/src/common/hooks/use_alerts_data_view.ts

```diff
@@ -0,0 +1,164 @@
+import { i18n } from '@kbn/i18n';
+import { DataViewsContract, FieldSpec } from '@kbn/data-views-plugin/common';
+import { AlertConsumers, ValidFeatureId } from '@kbn/rule-data-utils';
+import type { ToastsStart, HttpStart } from '@kbn/core/public';
+
+import { DataViewBase } from '@kbn/es-query';
+import { useVirtualDataViewQuery } from './use_virtual_data_view_query';
+import { useFetchAlertsFieldsQuery } from './use_fetch_alerts_fields_query';
+import { useFetchAlertsIndexNamesQuery } from './use_fetch_alerts_index_names_query';
+
+export interface UseAlertsDataViewParams {
+  // Dependencies
+  http: HttpStart;
+  dataViewsService: DataViewsContract;
+  toasts: ToastsStart;
+
+  // Params
+  /**
+   * Array of feature ids used for authorization and area-based filtering
+   *
+   * Security data views must be requested in isolation (i.e. `['siem']`). If mixed with
+   * other feature ids, the resulting data view will be empty.
+   */
+  featureIds: ValidFeatureId[];
+}
+
+export interface UseAlertsDataViewResult {
+  isLoading: boolean;
+  dataView?: Omit<DataViewBase, 'fields'> & { fields: FieldSpec[] };
+}
+
+/**
+ * Computes a {@link DataViewBase} object for alerts indices based on the provided feature ids
+ *
+ * @returns
+ * A {@link DataViewBase} object, intentionally not typed as a complete {@link DataView} object
+ * since only Security Solution uses an actual in-memory data view (when `featureIds = ['siem']).
+ * In all other cases the data view is computed from the index names and fields fetched from the
+ * alerting APIs.
+ */
+export const useAlertsDataView = ({
+  http,
+  dataViewsService,
+  toasts,
+  featureIds,
+}: UseAlertsDataViewParams): UseAlertsDataViewResult => {
+  const includesSecurity = featureIds.includes(AlertConsumers.SIEM);
+  const isOnlySecurity = featureIds.length === 1 && includesSecurity;
+
+  const {
+    data: indexNames,
+    isError: isIndexNamesError,
+    isLoading: isLoadingIndexNames,
+    isInitialLoading: isInitialLoadingIndexNames,
+  } = useFetchAlertsIndexNamesQuery(
+    { http, featureIds },
+    {
+      // Don't fetch index names when featureIds includes both Security Solution and other features
+      enabled: !!featureIds.length && (isOnlySecurity || !includesSecurity),
+    }
+  );
+
+  const {
+    data: fields,
+    isError: isFieldsError,
+    isLoading: isLoadingFields,
+    isInitialLoading: isInitialLoadingFields,
+  } = useFetchAlertsFieldsQuery(
+    { http, featureIds },
+    {
+      // Don't fetch fields when featureIds includes Security Solution
+      enabled: !!featureIds.length && !includesSecurity,
+    }
+  );
+
+  const { data: virtualDataView, isError: isVirtualDataViewError } = useVirtualDataViewQuery(
+    {
+      dataViewsService,
+      indexNames,
+    },
+    {
+      // Create data view only when featureIds = ['siem'] and indexNames are fetched
+      enabled: isOnlySecurity && !!indexNames?.length,
+    }
+  );
+
+  useEffect(() => {
```

Comment:
Using a `useEffect` here for 3 reasons:
1. we want to show only one error message for any failing request, which would require to extend all hooks to receive callback overrides,
2. `useQuery` callbacks are deprecated and [were removed in v5](https://tkdodo.eu/blog/breaking-react-querys-api-on-purpose),
3. while the article suggests to use global QueryClient callbacks, that would make nearly impossible to target this combined queries case

Reply:
Thanks for the explanation! I think it is ok for this PR to use the `useEffect`. Still, it can be called three times and show three toasters. Imagine each call failing with a difference of 1-2 seconds. Then the `useEffect` will be called one for the first call, then the second call will fail, then the `useEffect` will be called one for the second call, etc. At some point we would need to go with the global approach and figure out a way to handle our errors there by using the `meta` property.

Reply:
Oh sure, I guess the point here (same as it was before with the `onError` callbacks) was to show the same message without triplicating the same logic, not showing it only once. I'm afraid the global callbacks aren't deduplicated across different queries ([they run once _per query_](https://tkdodo.eu/blog/breaking-react-querys-api-on-purpose#:~:text=this%20callback%20will%20only%20be%20invoked%20once%20per%20query.)), but that shouldn't be too hard to implement once we decide the heuristics (i.e. account for retries, elapsed time from one error to another, etc.). We could build a utility for that! 🙂